x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/InferenceProcessor.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ml.inference;
+
+import org.elasticsearch.ingest.AbstractProcessor;
+import org.elasticsearch.ingest.ConfigurationUtils;
+import org.elasticsearch.ingest.IngestDocument;
+import org.elasticsearch.ingest.Processor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+
+/**
+ * An inference ingest processor.
+ * The processor performs inference on ingest documents using a pre-trained model.
+ * The type of the pre-trained model is defined via the processor config.
+ * Inference models implement {@link Model} and are loaded via a {@link ModelLoader}
+ */
+public class InferenceProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "inference";
+    private static final String MODEL_ID = "model_id";
+    private static final String MODEL_TYPE = "model_type";
+    private static final String IGNORE_MISSING = "ignore_missing";
+
+    private final Model model;
+
+    public InferenceProcessor(String tag, Model model) {
+        super(tag);
+        this.model = model;
+    }
+
+    @Override
+    public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Exception> handler) {
+        model.infer(ingestDocument, handler);
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) {
+        assert false : "The async override of execute() must be used";
```

Comment:
For my education: Why did you choose to `assert` rather than, say throw `UnsupportedOperationException`?

Reply:
No it should be `UnsupportedOperationException` thanks